firebase/test/core/data/gateways/filesystem-gateway.spec.ts

```diff
@@ -15,52 +15,50 @@ describe("FileSystemGateway", () => {
 
   const fakeFilesContents = ["file1Contents", "file2Contents"];
 
-  // TODO(matuella): How to mock while maintaining types?
-  // This way, we won't have to call `fsMock.expect("unsafeMethodCall")` and could
-  // possibly use `createSinonStub`
   let sandbox: sinon.SinonSandbox;
-  let fsMock: sinon.SinonMock;
+  let fsStub: sinon.SinonStubbedInstance<typeof fs.promises>;
 
   beforeEach(() => {
     sandbox = sinon.createSandbox();
-    fsMock = sandbox.mock(fs.promises);
+    fsStub = sandbox.stub(fs.promises);
     sandbox.stub(process, "cwd").returns(fakeCwd);
   });
 
   afterEach(() => {
     sandbox.restore();
+    sandbox.resetHistory();
   });
 
   it("should reject when readdir throws", async () => {
     const fsGateway = new FileSystemGateway();
-    fsMock.expects("readdir").rejects();
+    fsStub.readdir.rejects();
     await assert.rejects(() => fsGateway.readDirFilesAsStrings("any"), FilesystemError);
   });
 
   it("should reject when readFile throws", async () => {
     const fsGateway = new FileSystemGateway();
-    fsMock.expects("readdir").returns(fakeFiles);
-    fsMock.expects("readFile").rejects();
+    // Disable TS type-checking for `fsStub.readdir` call because `sinon` type definitions always invoke the last
+    // function signature by default when a function has overloads.
+    // See https://github.com/DefinitelyTyped/DefinitelyTyped/issues/36436 for more details about this issue.
```

Comment:
Could we add this to `test/README` instead? I think it's really relevant and should not live only here next to this comment.

Reply:
Sure